### PR TITLE
JS runnable browser

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR adds support for running the full browser (for chapters 2-8) in the browser, with demo browsers for chapters 2, 3, and 4. Besides some bug fixes, the main changes are:

1. Support for `bind` in the runtime. All the keys used in the chapters to date are implemented, and extending this shouldn't be too painful. This also required special-casing `bind` in the compiler to deal with "with".
2. Support for using `fetch` to fetch real web pages. The main change here is special-casing `makefile` to be an async function, so that it can make async requests. Naturally the web page needs to allow CORS to be fetchable.

### CODE REVIEW

File: www/widgets/rt.js [hunk 1]
```diff
@@ -99,7 +113,33 @@ static tkinter(options) {
         }
 
         bind(key, fn) {
-            // this.tk.addEventListener(...)
+            if (['<Up>', '<Down>', '<Left>', '<Right>'].indexOf(key) !== -1) {
```
Comment (chrishtr) [inline www/widgets/rt.js:116]: Is typing in the URL bar supposed to work? I tried building your branch just now and it didn't seem to.

File: www/widgets/lab2-browser.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<!doctype html>
```
Comment (chrishtr) [inline www/widgets/lab2-browser.html:1]: These files are all the same - use a pandoc template?

Comment (chrishtr) [inline www/widgets/lab2-browser.html:1]: These files are all the same - use a pandoc template?

File: infra/compile.py [hunk 3]
```diff
@@ -191,7 +191,13 @@ def load_outline(ol):
 def compile_method(base, name, args, ctx):
     base_js = compile_expr(base, ctx)
     args_js = [compile_expr(arg, ctx) for arg in args]
-    if name in LIBRARY_METHODS:
+    if name == "bind": # Needs special handling due to "this"
+        assert len(args) == 2
+        return base_js + ".bind(" + args_js[0] + ", (e) => " + args_js[1] + "(e))"
+    elif name == "makefile":
+        assert len(args) == 3
+        return "await " + base_js + ".makefile(" + ", ".join(args_js) + ")"
```
Comment (chrishtr) [inline infra/compile.py:199]: clever!

File: www/widgets/rt.js (hunk 1 above)
Comment (pavpanchekha) [reply www/widgets/rt.js:116]: It doesn't yet. I would like to add support, but I need to make a UI to explain that the web version is limited to same-origin.

File: www/widgets/lab2-browser.html (hunk 2 above)
Comment (pavpanchekha) [reply www/widgets/lab2-browser.html:1]: That's a good idea!

File: www/widgets/rt.js (hunk 1 above)
Comment (pavpanchekha) [reply www/widgets/rt.js:116]: URL is now configurable, but has to stay on the "browser.engineering" domain. Avoids having to explain this stuff.

File: www/widgets/lab2-browser.html (hunk 2 above)
Comment (pavpanchekha) [reply www/widgets/lab2-browser.html:1]: It templatized it, but we don't have an easy way to link chapter numbers (like `4`) to chapter names (like `html`) so all the templates right now point to the "graphics" chapter. Not ideal, I'll think about ways to fix it.

**Comment by pavpanchekha** [comment]
I believe this is ready to merge

**Review by chrishtr — APPROVED** [review]
LGTM % my comments from the other PR

**Pull request merged**